ethexe/consensus/src/validator/initial.rs

```diff
@@ -0,0 +1,242 @@
+                let producer = self.producer_for(block.header.timestamp, &data.validators);
+                let my_address = self.ctx.pub_key.to_address();
+
+                if my_address == producer {
+                    log::info!("👷 Start to work as a producer for block: {}", block.hash);
+
+                    Producer::create(self.ctx, block.clone(), data.validators)
+                } else {
+                    // TODO +_+_+: add test (in ethexe-service) for case where is not validator for current block
+                    let is_validator_for_current_block =
+                        data.validators.iter().any(|v| *v == my_address);
+
+                    log::info!(
+                        "👷 Start to work as a subordinate for block: {}, producer is {producer}, \
+                        I'm validator for current block: {is_validator_for_current_block}",
+                        block.hash
+                    );
+
+                    Subordinate::create(
+                        self.ctx,
+                        block.clone(),
+                        producer,
+                        is_validator_for_current_block,
+                    )
+                }
+            }
+            _ => DefaultProcessing::synced_block(self, data),
+        }
+    }
+}
+
+impl Initial {
+    pub fn create(ctx: ValidatorContext) -> Result<Box<dyn StateHandler>> {
+        Ok(Box::new(Self {
+            ctx,
+            state: State::WaitingForChainHead,
+        }))
+    }
+
+    // TODO #4555: block producer could be calculated right here, using propagation from previous blocks.
+    pub fn create_with_chain_head(
+        ctx: ValidatorContext,
+        block: SimpleBlockData,
+    ) -> Result<Box<dyn StateHandler>> {
+        Ok(Box::new(Self {
+            ctx,
+            state: State::WaitingForSyncedBlock(block),
+        }))
+    }
+
+    fn producer_for(&self, timestamp: u64, validators: &[Address]) -> Address {
+        let slot = timestamp / self.ctx.slot_duration.as_secs();
+        let index = crate::block_producer_index(validators.len(), slot);
+        validators
+            .get(index)
+            .cloned()
+            .unwrap_or_else(|| unreachable!("index must be valid"))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::any::TypeId;
+
+    use gprimitives::H256;
+
+    use super::*;
+    use crate::{tests::*, validator::tests::*, ConsensusEvent};
+
+    #[test]
+    fn create_initial_success() {
+        let (ctx, _) = mock_validator_context();
+        let initial = Initial::create(ctx).unwrap();
+        assert_eq!(initial.type_id(), TypeId::of::<Initial>());
+    }
+
+    #[test]
+    fn create_with_chain_head_success() {
+        let (ctx, _) = mock_validator_context();
+        let block = mock_simple_block_data();
+        let initial = Initial::create_with_chain_head(ctx, block.clone()).unwrap();
+        assert_eq!(initial.type_id(), TypeId::of::<Initial>());
+    }
+
+    #[tokio::test]
```

Comment:
why async

Reply:
you don't need async context if there's no async calls

Reply:
Discussed